designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,137 @@
+
+## Motivation
+
+A common use-case of the `warn` level is to report errors to the developer in their IDE,
+but to completely ignore it when running ESLint on the command line via the `--quiet` flag.
+When running with many rules set to the `warn` level, this can cause a significant amount
+of overhead on ESLint runtime. On an example large codebase with 154988 lines of TypeScript
+code, I found that by disabling all rules set to warn, I was able to reduce ESLint runtime
+from 4 minutes and 39 seconds to just 22 seconds. As the `warn` rules are entirely ignored
+in this situation, this is a substantial amount of wasted time and resources that could be
+avoided.
+
+## Detailed Design
+
+The proposed implementation is to modify the `--quiet` flag so that ESLint will skip
+running any rules that are set to the `warn` level.
+
+From an API perspective, this would be implemented by a predicate function that acts as
+a filter on which rules should be run. The predicate would take the rule configuration,
+and return whether the rule should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+In `cli.js`'s `translateOptions` function, the `rulePredicate` option should be assigned to
+a predicate that checks for a `severity` of 2 (error) when the `--quiet` flag is applied, otherwise
+always returns true. In `eslint/flat-eslint.js`, the `rulePredicate` should be taken from
+the `eslintOptions` object, and passed down to the `linter.verifyAndFix` call.
+
+Within `linter.js`, the API should be added to `VerifyOptions`, and will be passed down into and
+utilized within the `runRules` function during the `configuredRules` loop, after the check
+for disabled rules and the rule existing. The `ruleId`, `rule`, rule configuration from `getRuleOptions`,
+and `severity` should be passed into the predicate function as an object. `normalizeVerifyOptions`
+should verify that the `rulePredicate` option is a function, and replace it with an always-true
+function if not. `processOptions` in `eslint-helpers.js` should also perform a validation check
+that the `rulePredicate` option is a function.
+
+The alteration to the `--quiet` flag would be implemented by passing a predicate function
+to the API that only returns true for rules that have been set to `error`, thus filtering
+out any that are marked as `warn`.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--quiet`. As the
+rules are not actually run, an assumption would have to be made that all directives
+on rules marked `warn` are used when in this mode. This is a reasonable assumption, as
+the user likely does not expect `warn` flags to be touched at all in this mode.
```

Comment:
Can you provide more design details on this? This logic is inside the API, so it would use the predicate function rather than checking for the 'warn' severity?

Reply:
I've updated this to refer to any rules that return `false` on the predicate. Unsure if this is necessarily the best way to handle this but I feel there aren't really better alternatives.